### PR TITLE
Fixes seg prefix resolution when multiple prefixed dirs are present

### PR DESCRIPTION
**Pull request by soumyadeep2007**
Before this fix: GetSegPrefix would incorrectly match on the
directory lexicographically occurring first in order to compute the
prefix. Now, we also incorporate the timestamp into the resolution
process ensuring a correct match.

Authored-by: Soumyadeep Chakraborty <sochakraborty@pivotal.io>

### CODE REVIEW

File: utils/filepath.go [hunk 1]
```diff
@@ -159,15 +159,17 @@ func GetSegPrefix(connection *dbconn.DBConn) string {
...
+		masterDir, err := operating.System.Glob(fmt.Sprintf("%s/*-1/backups/*/%s", backupDir, timestamp))
 		if err != nil || len(masterDir) == 0 {
 			gplog.Fatal(err, "Master backup directory in %s missing or inaccessible", backupDir)
 		}
-		_, segPrefix = path.Split(masterDir[0])
-		segPrefix = segPrefix[:len(segPrefix)-2] // Remove "-1" segment ID from string
+		indexOfBackupsSubstr := strings.LastIndex(masterDir[0], "-1/backups/")
+		segPrefix = masterDir[0][:indexOfBackupsSubstr]
+		_, segPrefix = path.Split(segPrefix)
+		fmt.Println(segPrefix)
```
Comment (chrishajas) [inline utils/filepath.go]: Remove println

File: utils/filepath.go [hunk 2]
```diff
@@ -159,15 +159,17 @@ func GetSegPrefix(connection *dbconn.DBConn) string {
 	return segPrefix
 }
 
-func ParseSegPrefix(backupDir string) string {
+func ParseSegPrefix(backupDir string, timestamp string) string {
 	segPrefix := ""
 	if len(backupDir) > 0 {
-		masterDir, err := operating.System.Glob(fmt.Sprintf("%s/*-1", backupDir))
+		masterDir, err := operating.System.Glob(fmt.Sprintf("%s/*-1/backups/*/%s", backupDir, timestamp))
```
Comment (chrishajas) [inline utils/filepath.go]: This is no longer the masterDir, but rather the backup directory for the timestamp.

File: utils/filepath.go [hunk 3]
```diff
@@ -159,15 +159,17 @@ func GetSegPrefix(connection *dbconn.DBConn) string {
...
 		if err != nil || len(masterDir) == 0 {
 			gplog.Fatal(err, "Master backup directory in %s missing or inaccessible", backupDir)
 		}
-		_, segPrefix = path.Split(masterDir[0])
-		segPrefix = segPrefix[:len(segPrefix)-2] // Remove "-1" segment ID from string
+		indexOfBackupsSubstr := strings.LastIndex(masterDir[0], "-1/backups/")
+		segPrefix = masterDir[0][:indexOfBackupsSubstr]
```
Comment (chrishajas) [inline utils/filepath.go]: This isn't just the segPrefix right, does it also include the path up to the segPrefix?

Comment (soumyadeep2007) [reply utils/filepath.go]: After the split is executed, it is just the prefix.

File: utils/filepath.go (hunk 1 above)
Comment (soumyadeep2007) [reply utils/filepath.go]: Will do!

File: utils/filepath.go (hunk 2 above)
Comment (soumyadeep2007) [reply utils/filepath.go]: I was wondering what to call that! Thanks!

File: utils/filepath.go [hunk 4]
```diff
@@ -159,15 +159,15 @@ func GetSegPrefix(connection *dbconn.DBConn) string {
...
 	if len(backupDir) > 0 {
-		masterDir, err := operating.System.Glob(fmt.Sprintf("%s/*-1", backupDir))
-		if err != nil || len(masterDir) == 0 {
+		backupDirForTimestamp, err := operating.System.Glob(fmt.Sprintf("%s/*-1/backups/*/%s", backupDir, timestamp))
+		if err != nil || len(backupDirForTimestamp) == 0 {
 			gplog.Fatal(err, "Master backup directory in %s missing or inaccessible", backupDir)
 		}
-		_, segPrefix = path.Split(masterDir[0])
-		segPrefix = segPrefix[:len(segPrefix)-2] // Remove "-1" segment ID from string
+		indexOfBackupsSubstr := strings.LastIndex(backupDirForTimestamp[0], "-1/backups/")
```
Comment (khuddlefish) [inline utils/filepath.go]: Do we have any concern that we might get more than 1 result?

Comment (khuddlefish) [reply utils/filepath.go]: Also, does this get the index of "backups/" or "-1/backups/"?

Comment (soumyadeep2007) [reply utils/filepath.go]: Now we are searching with the timestamp in the dir path. So there can be only one directory captured in backupDirForTimestamp.

Gets last index of "-1/backups/".

**Comment by khuddlefish** [comment]
Do we have any tests that simulate the original issue we were seeing of getting multiple backup dirs so we can confirm that our original issue isn't present anymore?

**Comment by chrishajas** [comment]
I think most of the logic is in the glob'd result, which we mock.

**Pull request merged**